crates/matrix-sdk-indexeddb/src/stream.rs

```diff
@@ -0,0 +1,624 @@
+        let first_item = cursor.as_ref().and_then(|cursor| cursor.key().zip(Some(cursor.value())));
+
+        Self { cursor, continue_cursor_future: None, first_item }
+    }
+}
+
+impl<'a> Stream for StreamByCursor<'a> {
+    type Item = Result<(JsValue, JsValue), DomException>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.project();
+
+        // Let's get the cursor.
+        let Some(cursor) = &this.cursor else {
+            // No cursor? It means no matching records is found.
+            // There is no possible value. Let's close the `Stream`.
+            return Poll::Ready(None);
+        };
+
+        // A first result is ready. Let's return it.
+        // This operation happens once.
+        if let Some(value) = this.first_item.take() {
+            return Poll::Ready(Some(Ok(value)));
+        }
+
+        // The first value has been consumed. Let's poll for the next one?
+        //
+        // `continue_cursor` has not been called yet. Let's call it.
+        if this.continue_cursor_future.is_none() {
+            *this.continue_cursor_future = Some(Box::pin(cursor.continue_cursor()?));
+        }
+
+        // Now that `continue_cursor` has been called, we can poll its future.
+        //
+        // It returns a `bool`: `true` if a next value exists, `false` otherwise.
+        if ready!(this
+            .continue_cursor_future
+            .as_mut()
+            // SAFETY: The case where `continue_cursor_future` is `None` has been handled
+            // hereinabove.
+            .unwrap()
+            .as_mut()
+            .poll(cx))?
+        {
+            // At this point, we have a value ready to be returned. The
+            // `continue_cursor_future` can be dropped.
+            *this.continue_cursor_future = None;
+
+            Poll::Ready(Some(Ok((
+                cursor.key().unwrap_or_else(|| JsValue::UNDEFINED),
+                cursor.value(),
+            ))))
+        } else {
+            // It doesn't have a new value. End of the cursor. End of the stream.
+            Poll::Ready(None)
+        }
+    }
+}
+
+pin_project! {
+    /// `StreamByRenewedCursor` is an async iterator (a [`Stream`]) that traverses
+    /// an [`IdbObjectStore`] by using a cursor ([`IdbCursorWithValue`]) and a range
+    /// ([`IdbKeyRange`]).
+    ///
+    /// Why a range? What is renewed? Well, that's kind of complex. An
+    /// [`IdbTransaction`] is dropped by the runtime (most of the time, the
+    /// JavaScript runtime) when the transaction is considered “idle”. It's hard to
+    /// estimate when a transaction will be dropped, and it can happen in
+    /// circumstances out of control. One solution to this is to _renew_ the
+    /// transaction every time _x_ values are fetched from it. The lower _x_ is, the
+    /// higher the probability for the transaction to not be dropped is.
```

Comment:
So, based on my above comment, this is not about making dropping the transaction unlikely because less time passed: it's about having well-defined break points where we can stop streaming, write to disk, and then stream again.